### PR TITLE
Collision Detection & Response

### PR DESCRIPTION
**Pull request by Ruixel**
Octree generation is temporarily disabled until I find out why it's excluding overlapping walls

### CODE REVIEW

File: src/Editor/BoundingBox.cpp [hunk 1]
```diff
@@ -12,6 +12,37 @@ bool BoundingBox::checkAABB(const BoundingBox& obj)
...
+{
+	return (point.x > m_vecMin.x &&
+		point.x < m_vecMax.x &&
+		point.y > m_vecMin.y &&
+		point.y < m_vecMax.y &&
+		point.z > m_vecMin.z &&
+		point.z < m_vecMax.z);
+}
+
+glm::vec3 BoundingBox::getVecMin()
```
Comment (Hopson97) [inline src/Editor/BoundingBox.cpp]: could return const ref rather than copy

File: src/Editor/BoundingBox.cpp [hunk 2]
```diff
@@ -12,6 +12,37 @@ bool BoundingBox::checkAABB(const BoundingBox& obj)
...
+{
+	return m_vecMin;
+}
+
+glm::vec3 BoundingBox::getVecMax()
+{
+	return m_vecMax;
+}
```
Comment (Hopson97) [inline src/Editor/BoundingBox.cpp]: same here, could return const ref rather than copy

File: src/Editor/BoundingBox.h [hunk 3]
```diff
@@ -5,6 +5,7 @@
 #include "../Renderer/Renderer.h"
 
 #include "glm/vec3.hpp"
+#include "../Game/WorldConstants.h"
 
 #define RENDER_BOUNDING_BOX true
```
Comment (Hopson97) [inline src/Editor/BoundingBox.h]: https://github.com/ChallengeYou-Resurrection/Plan-and-Info/blob/master/CppCodestyle.md#constants

File: src/Editor/BoundingBox.cpp [hunk 4]
```diff
@@ -12,6 +12,37 @@ bool BoundingBox::checkAABB(const BoundingBox& obj)
...
+{
+	return (this->m_vecMax.x > obj_minv3.x &&
+		this->m_vecMin.x < obj_maxv3.x &&
+		this->m_vecMax.y > obj_minv3.y &&
+		this->m_vecMin.y < obj_maxv3.y &&
+		this->m_vecMax.z > obj_minv3.z &&
+		this->m_vecMin.z < obj_maxv3.z);
+}
+
+bool BoundingBox::checkAABB(const glm::vec3 & point)
```
Comment (Hopson97) [inline src/Editor/BoundingBox.cpp]: function does not affect the class, so should be marked as `const`

File: src/Editor/GeoOctree.cpp [hunk 5]
```diff
@@ -1,38 +1,116 @@
...
+			m_subdivisionBB.at(i).getVecMin());
+
+	// Success
+	return true;
 }
 
-/*
-void GeoOctree::insertGeometry(std::unique_ptr<EditorObject> obj)
+bool GeoOctree::checkIfWallInsideAABB(const Wall& wall, const glm::vec3& min,
+	const glm::vec3& max)
```
Comment (Hopson97) [inline src/Editor/GeoOctree.cpp]: class not changed by the function, so can be marked as a `const` function

File: src/Editor/GeoOctree.cpp [hunk 6]
```diff
@@ -41,7 +119,54 @@ void GeoOctree::cleanOctree() {
...
+
 	// Draw bounding boxes as wireframe
-	for (auto& box : m_subdivisionBB)
-		box.render(renderer);
+	if (subdivided)
+		for (auto& node : m_nodes)
+			node->drawOctree(renderer);
+}
+
+int GeoOctree::getObjectSize()
```
Comment (Hopson97) [inline src/Editor/GeoOctree.cpp]: const function

File: src/Editor/GeoOctree.cpp [hunk 7]
```diff
@@ -41,7 +119,54 @@ void GeoOctree::cleanOctree() {
...
+	for (auto& node : m_nodes)
+		if (node->checkPointInOctree(point))
+			return node->getWallVectorNearPoint(point);
+
+	// fail safe (Caused by point outside of the octree)
+	//std::cout << "Warning: Point is outside the octree\n";
+	return std::vector<std::shared_ptr<Wall>>();
 }
+
+bool GeoOctree::checkPointInOctree(const glm::vec3 & point)
```
Comment (Hopson97) [inline src/Editor/GeoOctree.cpp]: const function

File: src/Game/MeshBuilder.h [hunk 8]
```diff
@@ -2,7 +2,40 @@
...
 #include "WorldTextures.h"
+#include "CYObjects.h"
 
 struct Wall;
 
-Mesh createMesh(const Wall& wall, const WorldTextures& wTex);
\ No newline at end of file
+Mesh createMesh(const Wall& wall, const WorldTextures& wTex);
+
+namespace {
```
Comment (Hopson97) [inline src/Game/MeshBuilder.h]: why was this moved to the header file?

**Review by Hopson97 — CHANGES_REQUESTED** [review]

File: src/Game/MeshBuilder.h (hunk 8 above)
Comment (Ruixel) [reply src/Game/MeshBuilder.h]: I was having linker issues at the time. I put it back into the .cpp and used a namespace
